Read and use form definition for parameters if supplied by broker.

cfchase opened the pull request:
Allows the OpenShift UI to enforce order, grouping, and display type, if that information is supplied by the broker.  Brokers are responsible for adding the form definition in the plan metadata as `schemas.service_instance.create.form_definition`  (A similar path for service_binding.create can also be supplied this way, but out of scope fort his PR).   

Unfortunately, regarding grouping, fieldset and section html is generated properly, but has no visual difference.  This would take additional work to style appropriately. (picture below)

Example broker code:
https://github.com/openshift/ansible-service-broker/pull/386 

Example json gist:
https://gist.github.com/cfchase/8614236dcb8e22cc0aa8a2b4f1ee52d0

Screenshots
Full Wizard:
![wizard-full](https://user-images.githubusercontent.com/1448375/29636810-0e7a893e-8820-11e7-9017-f7461ef7021f.png)

Example display types of fields:
![wizard-fields-1](https://user-images.githubusercontent.com/1448375/29636813-12d3737e-8820-11e7-8ab5-e1831076e0ab.png)
![wizard-fields-2](https://user-images.githubusercontent.com/1448375/29636817-160ad942-8820-11e7-9bbd-ad4efd251c9a.png)
![wizard-fields-3](https://user-images.githubusercontent.com/1448375/29636821-17ea7a10-8820-11e7-9f91-12747fa8308b.png)

Fieldset not styled:
![wizard-fieldset-no-style](https://user-images.githubusercontent.com/1448375/29636906-5f96d872-8820-11e7-91d5-338e106f7ecd.png)

cfchase commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(8) Parameter Ordering and Grouping [8]](https://trello.com/c/Orp4rGpI/441-8-parameter-ordering-and-grouping-8)

spadgett commented on src/components/order-service/order-service.controller.ts:
```diff
@@ -312,6 +313,34 @@ export class OrderServiceController implements angular.IController {
+    this.ctrl.parameterForm = formDefinition || ['*'];
+  }
+
+  private removeUnsupportedParameterFormItems(formItems: any) {
```
It would be better to move this in the catalog-parameters component so that we will be able to reuse it for bind parameters.

spadgett commented on src/components/order-service/order-service.controller.ts:
```diff
@@ -312,6 +313,34 @@ export class OrderServiceController implements angular.IController {
+      if (_.isObject(item)) {
+        _.forOwn(item, (value, key) => {
+          if (!allowedKeys[key]) {
+            delete item[key];
```
We should avoid mutating the original API object since it's cached by our DataService and possibly used outside of this controller. We should only modify a copy.

spadgett commented on src/components/order-service/order-service.controller.ts:
```diff
@@ -312,6 +313,34 @@ export class OrderServiceController implements angular.IController {
+      'items': true
+    };
+
+    _.forEach(formItems, (item: any) => {
```
Nit: we've been trying to consistently use `_.each` so we don't have a mix

spadgett commented on src/components/order-service/order-service.controller.ts:
```diff
@@ -312,6 +313,34 @@ export class OrderServiceController implements angular.IController {
+  private removeUnsupportedParameterFormItems(formItems: any) {
+    const allowedKeys = {
+      'key': true,
+      'type': true,
```
I'd like to lock down the possible values for type as well. I think this would be a good set:

- fieldset
- text
- textarea
- password
- checkbox
- select

https://github.com/json-schema-form/angular-schema-form/blob/development/docs/index.md#form-types

We might need to move away from angular-schema-form someday, and this list seems manageable if we ever need to implement it ourselves.

spadgett commented on src/components/order-service/order-service.controller.ts:
```diff
@@ -312,6 +313,34 @@ export class OrderServiceController implements angular.IController {
   }
 
+  private updateParameterForm() {
+    let formDefinition: any = _.get(this, 'ctrl.selectedPlan.externalMetadata.schemas.service_instance.create.form_definition');
```
Let's add a prefix to avoid ever colliding with a future field added to OSB API. I'd suggest `openshift_form_definition`. That also makes it clear to anyone looking at the broker that the field is specific to the OpenShift catalog.

@pmorie Any objections?

cfchase commented:
* changed field name to openshift_form_definition
* added comment back in with slight change
* changed deletion if bad keys to copying of object only with good keys and type values
* use _.each instead of _.forEach

@spadgett let me know if this addresses your comments or if there's more to do.

spadgett commented on src/components/catalog-parameters/catalog-parameters.controller.ts:
```diff
@@ -28,4 +39,39 @@ export class CatalogParametersController implements angular.IController {
+    let newValue: any;
+
+    if (_.isString(original)) {
+      newValue = original;
```
I think early returns would make this easier to follow:

```js
if (_.isString(original)) {
  return original;
}

// ...
```

spadgett commented on src/components/catalog-parameters/catalog-parameters.controller.ts:
```diff
@@ -28,4 +39,39 @@ export class CatalogParametersController implements angular.IController {
+        if (newItem) {
+          newValue.push(newItem);
+        }
+      });
```
`_.map` should help here

```js
if (_.isArray(original)) {
  return _.map(original, _.bind(this.cloneParameterForm, this));
}
```

cfchase commented:
@spadgett let me know how this styling looks to you or if it needs less spacing, etc.

* Cleaned up cloneParameterForm
* Added fieldset titles and styling
* Added mock data service for APB.

Screenshots of grouping

![wizard-full-groups](https://user-images.githubusercontent.com/1448375/29788843-0a1f2706-8c02-11e7-804a-0e470e288147.png)

![wizard-groups-1](https://user-images.githubusercontent.com/1448375/29788887-3518a02c-8c02-11e7-885e-88951efde242.png)
![wizard-groups-2](https://user-images.githubusercontent.com/1448375/29788850-10560df6-8c02-11e7-8341-b116161c4717.png)
![wizard-groups-3](https://user-images.githubusercontent.com/1448375/29788855-1240e370-8c02-11e7-838a-3429ec97d67c.png)

spadgett commented on src/styles/order-service.less:
```diff
@@ -203,4 +203,16 @@
+      padding-top: 20px;
+      margin-bottom: 10px;
+      border-top: solid 1px rgba(0, 0, 0, 0.15);
+      border-bottom: none;
```
nit: we try to keep the CSS styles alphabetized
